locator/token_metadata.cc

```diff
@@ -851,6 +862,14 @@ future<> token_metadata_impl::update_pending_ranges(
                 if (!pending_endpoints.empty()) {
                     add_mapping(migration_info.pending_endpoints, std::move(pending_endpoints));
                 }
+
+                // in order not to waste memory, we update read_endpoints only if the
+                // new endpoints differs from the old one
+                if (_topology_transition_state == service::topology::transition_state::write_both_read_new &&
```

Comment:
This check was in the different place in the previous version, so if such a state was added the test would have needed a fix anyway. The way to handle it in a more general way is to add `is_read_new_state(state&)` function into `service::topology` and still the one who adds a state will have to remember to update it.

Reply:
> The way to handle it in a more general way is to add is_read_new_state(state&) function into service::topology and still the one who adds a state will have to remember to update it.

this feels like a hacky way to say the same thing as in my prev version: just convert one set of states to another

Reply:
> > both sets are describing topology transitions.
> 
> They are not: the enum in `token_metadata` described just how reads should be conducted, not all sm states. It was meant to be equivalent to `bool read_new`, just more type safe, like `bool_class`. Maybe it should have been just `bool_class<class read_new_tag>`.

Exactly. If that was your intention `bool_class` is the one you should have used. But I actually liked setting the state, just did not like having additional enum to map to. More flexible IMO. But yes, there is a chance we will not need the flexibility.

Reply:
A way to protect against condition rotting in case we add more states is to have a switch on the state which causes compilation error until someone makes a decision how to handle the new state in this conversion.